### PR TITLE
Don't expect `application/json` type nodes to be files

### LINKED ISSUE #8538: `gatsby-transformer-json` incompatible with `gatsby-source-contentful`
## Description
I believe there's an incompatibility between `gatsby-transformer-json` and `gatsby-source-contentful` that results in an error on boot. I am experiencing this issue identically to #5680 (closed, marked stale). 

In short, `gatsby-transformer-json/gatsby-node.js` attempts to process JSON coming back from `gatsby-source-contentful` and fails when it's unable to determine the directory that the JSON is in. 

### Steps to reproduce

1. Open up the [Contentful example app](https://github.com/gatsbyjs/gatsby/tree/master/examples/using-contentful)
2. Install dependencies (`yarn`) and run develop
3. Confirm it works
4. `yarn add gatsby-transformer-json`
5. Add `gatsby-transformer-json` to `gatsby-config.js`
6. Attempt to run develop

### Expected result

The app behaves identically.

### Actual result

```
success open and validate gatsby-config — 0.009 s
success load plugins — 0.140 s
success onPreInit — 0.365 s
success delete html and css files from previous builds — 0.010 s
info One or more of your plugins have changed since the last time you ran Gatsby. As
a precaution, we're deleting your site's cache to ensure there's not any stale
data
success initialize cache — 0.015 s
success copy gatsby files — 0.025 s
success onPreBootstrap — 0.003 s
⠁ Starting to fetch data from Contentful
Fetching default locale
⠠ source and transform nodesdefault locale is : en-US
⠂ source and transform nodescontentTypes fetched 6
Updated entries  12
Deleted entries  0
Updated assets  11
Deleted assets  0
Fetch Contentful data: 558.445ms
error Plugin gatsby-transformer-json returned an error

  Error: TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Rece  ived type undefined
  
  - gatsby-node.js:40 getType
    [using-contentful]/[gatsby-transformer-json]/gatsby-node.js:40:49
  
  - gatsby-node.js:82 Object.<anonymous>
    [using-contentful]/[gatsby-transformer-json]/gatsby-node.js:82:113
  
  - Generator.next
  
  - util.js:16 tryCatcher
    [global]/[bluebird]/js/release/util.js:16:23

error Plugin gatsby-transformer-json returned an error

  Error: TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Rece  ived type undefined
  
  - gatsby-node.js:40 getType
    [using-contentful]/[gatsby-transformer-json]/gatsby-node.js:40:49
  
  - gatsby-node.js:82 Object.<anonymous>
    [using-contentful]/[gatsby-transformer-json]/gatsby-node.js:82:113
  
  - Generator.next
  
  - util.js:16 tryCatcher
    [global]/[bluebird]/js/release/util.js:16:23
  
error UNHANDLED REJECTION

  TypeError: Cannot read property 'filter' of undefined
  
  - api-runner-node.js:274 Promise.mapSeries.catch.then.results
    [using-contentful]/[gatsby]/dist/utils/api-runner-node.js:274:42
  
  - util.js:16 tryCatcher
    [using-contentful]/[bluebird]/js/release/util.js:16:23
```

(note: stacks truncated)

### Environment

  System:
    OS: macOS High Sierra 10.13.6
  Binaries:
    Node: 10.9.0 - /usr/local/bin/node
    Yarn: 1.9.4 - /usr/local/bin/yarn
    npm: 6.2.0 - /usr/local/bin/npm
  npmPackages:
    gatsby: ^2.0.0 => 2.0.8 
    gatsby-image: ^2.0.5 => 2.0.9 
    gatsby-plugin-google-analytics: ^2.0.5 => 2.0.6 
    gatsby-plugin-offline: ^2.0.5 => 2.0.5 
    gatsby-plugin-typography: ^2.2.0 => 2.2.0 
    gatsby-source-contentful: ^2.0.1 => 2.0.1 
    gatsby-transformer-json: ^2.1.1 => 2.1.1 
    gatsby-transformer-remark: ^2.1.1 => 2.1.3

### PR DESCRIPTION
**Pull request by coreyward**
`gatsby-source-contentful` creates nodes with a JSON type so we can't expect all nodes to have a `node.dir` property to generate a GraphQL type from. 

Fixes #8538.

### CODE REVIEW

File: packages/gatsby-transformer-json/src/gatsby-node.js [hunk 1]
```diff
@@ -10,6 +10,8 @@ async function onCreateNode({ node, actions, loadNodeContent, createNodeId }, pl
       return pluginOptions.typeName
     } else if (isArray) {
       return _.upperFirst(_.camelCase(`${node.name} Json`))
+    } else if (node.internal.type !== `File`) {
+      return `${node.internal.type} Json`
```
Comment (eLod) [inline packages/gatsby-transformer-json/src/gatsby-node.js]: i am not sure if this is the better option, as this dictates the graphql schema, e.g. the contentful sets `internal.type` to ``_.camelCase(`${node.internal.type} ${key} JSONNode`)`` so you would need to query by that name. do you have examples from contentful? what about `node.name`?

**Comment by eLod** [comment]
cc @pieh

File: packages/gatsby-transformer-json/src/gatsby-node.js (hunk 1 above)
Comment (coreyward) [reply packages/gatsby-transformer-json/src/gatsby-node.js]: Is `node.name` always available and suitable for GraphQL? I have no dog in the fight, just trying to get the incompatibility resolved so I can keep moving on a project.

Comment (pieh) [reply packages/gatsby-transformer-json/src/gatsby-node.js]: `node.name` is also `File` specific - but at least it won't crash if `name` is not set, but yeah, let's move this `else if` before `else if (isArray)`

**Comment by coreyward** [comment]
@pieh Thanks for the assistance, and for tests!

**Pull request merged**